code.js

```diff
@@ -1,3 +1,59 @@
 function tsp_hk(distance_matrix) {
-    return -1;
+    if (distance_matrix.length == 0 || distance_matrix.length == 1) {
+        return 0;
+    }
+
+    const cities = new Set(); //cities (including start) not visited yet
+    for (let city = 0; city < distance_matrix.length; city++) {
+        cities.add(city);
+    }
+    console.log(cities);
+    const start = 0;
+    const cache = new Map();
+    console.log("cache = ", cache);
+    
+    return heldKarp(distance_matrix, cities, start, cache);
+}
+
+function heldKarp(distance_matrix, cities, start, cache) {
+    if (Math.abs(cities.size) == 2) { //base case
+        console.log("2 cities in set");
+        for (let city of cities) {
+            console.log("city = ", city);
+            if (city != start) {
+                console.log("returning ", distance_matrix[start][city]);
+                return distance_matrix[start][city];
+            }
+        }
+    }
+    
+    let minDist = Infinity;
+    for (city of cities) {
```

Comment:
You're missing the scope for `city` here -- it defaults to `var`, which makes it a global variable.